### PR TITLE
✨ (config-gen alpha command): upgrade sigs.k8s.io/controller-tools from v0.7.0 to v0.8.0

### LINKED ISSUE #2567: Unit tests are not working locally (go 1.17.5)
### What broke? What's expected?

The make test is not working locally
That fails when we run `go test -race ./pkg/... `

### Reproducing this issue

It was reported on the channel that someone was not able to run the tests.
the tests are called in the CI and works see: https://github.com/kubernetes-sigs/kubebuilder/runs/5605172452?check_suite_focus=true#step:5:1

However, I tried to run with;
- Mac os (darwin_amd64)
- Go 1.17.5

And also faced issues: 

```
--- FAIL: TestNewCommand (0.03s)
    --- FAIL: TestNewCommand/testdata/componentconfig (0.02s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4e258ea]

goroutine 23 [running]:
testing.tRunner.func1.2({0x4f3a160, 0x59b1970})
	/usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1209 +0x36c
testing.tRunner.func1()
	/usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1212 +0x3b6
panic({0x4f3a160, 0x59b1970})
	/usr/local/Cellar/go/1.17.5/libexec/src/runtime/panic.go:1047 +0x266
sigs.k8s.io/kubebuilder/v3/pkg/cli/alpha/config-gen.ControllerGenFilter.Filter({0x0}, {0x0, 0x0, 0x4071325})
	/Users/camilamacedo86/go/src/sigs.k8s.io/kubebuilder/pkg/cli/alpha/config-gen/controller-gen-filter.go:68 +0x58a
sigs.k8s.io/kustomize/kyaml/fn/framework.(*TemplateProcessor).doPreProcess(0xc00016cfc0, {0x0, 0x0, 0x0})
	/Users/camilamacedo86/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.21/fn/framework/processors.go:280 +0x17b
sigs.k8s.io/kustomize/kyaml/kio.FilterFunc.Filter(0xc000427ae0, {0x0, 0x0, 0x0})
	/Users/camilamacedo86/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.21/kio/kio.go:60 +0x52
sigs.k8s.io/kustomize/kyaml/kio.Pipeline.ExecuteWithCallback({{0xc0004a5178, 0x1, 0x1}, {0xc0004a51f0, 0x5, 0x5}, {0xc0004a5168, 0x1, 0x1}, 0x1}, ...)
	/Users/camilamacedo86/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.21/kio/kio.go:120 +0x32b
sigs.k8s.io/kustomize/kyaml/kio.Pipeline.Execute(...)
	/Users/camilamacedo86/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.21/kio/kio.go:93
sigs.k8s.io/kustomize/kyaml/fn/framework.TemplateProcessor.Filter({{0x5004bc0, 0xc0001bcc00}, {0xc0003cdcc0, 0x1, 0x1}, {0xc000473710, 0x3, 0x3}, 0x1, {0xc0003cdca0, ...}, ...}, ...)
	/Users/camilamacedo86/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.21/fn/framework/processors.go:245 +0x73a
sigs.k8s.io/kustomize/kyaml/fn/framework.(*ResourceList).Filter(0xc000473b00, {0x5239740, 0xc00016cf30})
	/Users/camilamacedo86/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.21/fn/framework/framework.go:144 +0x74
sigs.k8s.io/kustomize/kyaml/fn/framework.TemplateProcessor.Process({{0x5004bc0, 0xc0001bcc00}, {0xc0003cdcc0, 0x1, 0x1}, {0xc000473710, 0x3, 0x3}, 0x1, {0xc0003cdca0, ...}, ...}, ...)
	/Users/camilamacedo86/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.21/fn/framework/processors.go:259 +0x138
sigs.k8s.io/kustomize/kyaml/fn/framework.Execute({0x5239720, 0xc00016ccf0}, 0xc000195970)
	/Users/camilamacedo86/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.21/fn/framework/framework.go:118 +0x39e
sigs.k8s.io/kustomize/kyaml/fn/framework/command.Build.func1(0xc00047a280, {0xc000427530, 0x1, 0x1})
	/Users/camilamacedo86/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.21/fn/framework/command/command.go:101 +0x40b
github.com/spf13/cobra.(*Command).execute(0xc00047a280, {0xc0004273b0, 0x1, 0x1})
	/Users/camilamacedo86/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:856 +0xa7e
github.com/spf13/cobra.(*Command).ExecuteC(0xc00047a280)
	/Users/camilamacedo86/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:974 +0x5db
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/camilamacedo86/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:902
sigs.k8s.io/kustomize/kyaml/fn/framework/frameworktestutil.(*CommandResultsChecker).Assert.func1()
	/Users/camilamacedo86/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.21/fn/framework/frameworktestutil/frameworktestutil.go:114 +0x505
sigs.k8s.io/kustomize/kyaml/fn/framework/frameworktestutil.(*resultsChecker).runDirectoryTestCase.func2(0xc000423d40)
	/Users/camilamacedo86/go/pkg/mod/sigs.k8s.io/kustomize/kyaml@v0.10.21/fn/framework/frameworktestutil/frameworktestutil.go:303 +0xe2
testing.tRunner(0xc000423d40, 0xc0004800c0)
	/usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1259 +0x230
created by testing.(*T).Run
	/usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1306 +0x727

```

then, we need to check why to solve it or add the guidance in the Contribution doc

### KubeBuilder (CLI) Version

master

### PROJECT version

_No response_

### Plugin versions

_No response_

### Other versions

_No response_

### Extra Labels

/kind regression

### PR DESCRIPTION
**Pull request by AlmogBaku**
## Description:
Update the controller-tools dep from 0.7.0 to .0.8.0 in the Kubebuilder CLI go mod which is only used by config-gen (alpha command). 

## Motivation
We could check that when we try to use the go version 1.18 the following errors are faced when the unit tests to check config-gen are executed. see :
```
/Users/AlmogBaku/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/builder/webhook.go:27:2: missing go.sum entry for module providing package k8s.io/client-go/rest (imported by sigs.k8s.io/controller-runtime/pkg/builder); to add:
	go get sigs.k8s.io/controller-runtime/pkg/builder@v0.11.2
/Users/AlmogBaku/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/client/patch.go:22:2: missing go.sum entry for module providing package github.com/evanphx/json-patch (imported by sigs.k8s.io/controller-runtime/pkg/client); to add:
	go get sigs.k8s.io/controller-runtime/pkg/client@v0.11.2
/Users/AlmogBaku/go/pkg/mod/k8s.io/kube-openapi@v0.0.0-20211115234752-e816edb12b65/pkg/util/proto/document.go:24:2: missing go.sum entry for module providing package github.com/googleapis/gnostic/openapiv2 (imported by k8s.io/kube-openapi/pkg/util/proto); to add:
	go get k8s.io/kube-openapi/pkg/util/proto@v0.0.0-20211115234752-e816edb12b65
/Users/AlmogBaku/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/client/watch.go:26:2: missing go.sum entry for module providing package k8s.io/client-go/dynamic (imported by sigs.k8s.io/controller-runtime/pkg/client); to add:
	go get sigs.k8s.io/controller-runtime/pkg/client@v0.11.2
```

## Why?

Tests were failing due to a missing `go.mod` project.
This was caused by a bizarre bug in `go list`, which it couldn't locate the appropriate go module.

The issue caused a lot of silent errors which complained upon the issue described above. Up until go 1.18 it was probably silenced totally, but now it's just panicing.

## Solution:
1. Do not ignore errors when creating the Runtime :)
2. Solve the root cause by adding a go module in the `pkg/cli/alpha/config-gen/testdata`
3. Upgrade outdated dependencies

## Notes:
This PR is high-priority since it's blocking other PRs
Closes #2567 

That was a tricky one ;-)

### CODE REVIEW

**Comment by camilamacedo86** [comment]
It is broken the unit-test checks
Could you please check?

File: go.mod [hunk 1]
```diff
@@ -15,14 +15,17 @@ require (
 	golang.org/x/tools v0.1.6-0.20210820212750-d4cc65f0b2ff
```
Comment (camilamacedo86) [inline go.mod:15]: This scenario does not solve the case: https://github.com/kubernetes-sigs/kubebuilder/pull/2581

Comment (camilamacedo86) [inline go.mod:15]: This scenario does not solve the case: https://github.com/kubernetes-sigs/kubebuilder/pull/2581

File: go.mod [hunk 2]
```diff
@@ -15,14 +15,17 @@ require (
...
 	k8s.io/apimachinery v0.23.5 // for `kubebuilder alpha config-gen`
 	sigs.k8s.io/controller-runtime v0.11.2
-	sigs.k8s.io/controller-tools v0.7.0 // for `kubebuilder alpha config-gen`
-	sigs.k8s.io/kustomize/kyaml v0.10.21 // for `kubebuilder alpha config-gen`
+	sigs.k8s.io/controller-tools v0.8.0 // for `kubebuilder alpha config-gen`
```
Comment (camilamacedo86) [inline go.mod:18]: We are actually here upgrading the `sigs.k8s.io/controller-tools` version used by the config gen so we ought to update the pr title for 

> ✨ (config-gen alpha command): upgrade sigs.k8s.io/controller-tools from v0.7.0 to v0.8.0

Comment (camilamacedo86) [inline go.mod:18]: We are actually here upgrading the `sigs.k8s.io/controller-tools` version used by the config gen so we ought to update the pr title for 

> ✨ (config-gen alpha command): upgrade sigs.k8s.io/controller-tools from v0.7.0 to v0.8.0

File: go.mod [hunk 3]
```diff
@@ -15,14 +15,17 @@ require (
...
+	sigs.k8s.io/kustomize/kyaml v0.13.6 // for `kubebuilder alpha config-gen`
 	sigs.k8s.io/yaml v1.3.0
 )
 
+require golang.org/x/text v0.3.7
```
Comment (camilamacedo86) [inline go.mod:23]:
```suggestion
```
```suggestion
 It does not seem required. See that this one will be the version imported by default : 
`golang.org/x/text v0.3.7 // indirect`

Comment (camilamacedo86) [inline go.mod:23]:
```suggestion
```
```suggestion
 It does not seem required. See that this one will be the version imported by default : 
`golang.org/x/text v0.3.7 // indirect`

Comment (camilamacedo86) [reply go.mod:23]: Could you please remove the required

Comment (camilamacedo86) [reply go.mod:23]: Could you please remove the required

File: pkg/cli/alpha/config-gen/controller-gen-filter.go [hunk 4]
```diff
@@ -64,7 +64,10 @@ func (cgr ControllerGenFilter) Filter(input []*yaml.RNode) ([]*yaml.RNode, error
 
 	// set the directory
 	b := bufferedGenerator{}
-	rt, _ := gens.ForRoots(cgr.Spec.CRDs.SourceDirectory) // ignore the spurious error
+	rt, err := gens.ForRoots(cgr.Spec.CRDs.SourceDirectory)
+	if err != nil {
+		return nil, errors.WrapPrefixf(err, "failed to produce generator's Runtime against %s", cgr.Spec.CRDs.SourceDirectory)
+	}
```
Comment (camilamacedo86) [inline pkg/cli/alpha/config-gen/controller-gen-filter.go:70]: @pwittrock could you please give a hand here? Why have we are ignoring these errors before?

Comment (camilamacedo86) [inline pkg/cli/alpha/config-gen/controller-gen-filter.go:70]: @pwittrock could you please give a hand here? Why have we are ignoring these errors before?

**Review by camilamacedo86 — COMMENTED** [review]
<removed because it is sorted out>

File: pkg/cli/alpha/config-gen/controller-gen-filter.go (hunk 4 above)
Comment (camilamacedo86) [inline pkg/cli/alpha/config-gen/controller-gen-filter.go:70]: Begin to return the error does not seems to be related to any go get issue.
That seems be solved only by the fact of we upgrade the controller-tools used by config-gen

I'd recommend:
- Revert this part and leave this PR just to upgrade the dep
- then we can check why the errors are ignored and if we should or not return them. That shows very intentional the reason for them be ignored.

File: go.mod (hunk 1 above)
Comment (AlmogBaku) [reply go.mod:15]: With this version, it does work with my environment with golang 1.17 + mac.
Can you please elaborate?

File: go.mod (hunk 3 above)
Comment (AlmogBaku) [reply go.mod:23]: `go mod tidy` add this... 🤷🏽‍♂️

File: go.mod (hunk 1 above)
Comment (camilamacedo86) [reply go.mod:15]: The problem faced/reported on the issue https://github.com/kubernetes-sigs/kubebuilder/pull/2581 is not solved with this PR.

**Comment by camilamacedo86** [comment]
@AlmogBaku, 

All test works so that is not a fix to them.
They broke when you try to bump the go 1.18
So, in order to be able to work with go 1.18 it shows that we need to upgrade the controller-tools version dep
which is used by config-gen. 

In this way, could you please update the title/description just so we have it clarified?

File: go.mod (hunk 3 above)
Comment (camilamacedo86) [reply go.mod:23]: OK. thank you for let me know

File: go.mod (hunk 1 above)
Comment (AlmogBaku) [reply go.mod:15]: Did you try it?
Because preventing the error provides a non nil Runtime(which should fix the bug)

Comment (AlmogBaku) [reply go.mod:15]: PS that is why I added the error returning (which you asked me to revert)

**Review by camilamacedo86 — APPROVED** [review]
@AlmogBaku , thank you for the great contribution 🥇 

Just a nit: The title description shows not accurate
Why?

> 🌱 fix unit tests (config-gen bug)

We have no bug in the tests. They work. 
The problem is faced only when we try to use go 1.18
So, we need to update the required deps used by config-gen 

Then, could we keep only the text from the :sparkles: ? 

Also, about the description? Could we update for something like follows?WDYT?

```
**Description**
Update the controller-tools dep from 0.7.0 to .0.8.0 in the Kubebuilder CLI go mod which is only used by config-gen (alpha command). 

**Motivation**
We could check that when we try to use the go version 1.18 the following errors are faced when the unit tests to check config-gen are executed. see :

```
```
/Users/AlmogBaku/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/builder/webhook.go:27:2: missing go.sum entry for module providing package k8s.io/client-go/rest (imported by sigs.k8s.io/controller-runtime/pkg/builder); to add:
	go get sigs.k8s.io/controller-runtime/pkg/builder@v0.11.2
/Users/AlmogBaku/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/client/patch.go:22:2: missing go.sum entry for module providing package github.com/evanphx/json-patch (imported by sigs.k8s.io/controller-runtime/pkg/client); to add:
	go get sigs.k8s.io/controller-runtime/pkg/client@v0.11.2
/Users/AlmogBaku/go/pkg/mod/k8s.io/kube-openapi@v0.0.0-20211115234752-e816edb12b65/pkg/util/proto/document.go:24:2: missing go.sum entry for module providing package github.com/googleapis/gnostic/openapiv2 (imported by k8s.io/kube-openapi/pkg/util/proto); to add:
	go get k8s.io/kube-openapi/pkg/util/proto@v0.0.0-20211115234752-e816edb12b65
/Users/AlmogBaku/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/client/watch.go:26:2: missing go.sum entry for module providing package k8s.io/client-go/dynamic (imported by sigs.k8s.io/controller-runtime/pkg/client); to add:
	go get sigs.k8s.io/controller-runtime/pkg/client@v0.11.2
```

Otherwise, 
/approved

**Comment by camilamacedo86** [comment]
The API diff fails here because of the changes in the yaml dep:

```
sigs.k8s.io/kustomize/kyaml/fn/framework
  Incompatible changes:
  - Field.CurrentValue: changed from string to interface{}
  - Field.SuggestedValue: removed
  - ResourceList.Result: removed
  - Result.Error: removed
  - Result.ExitCode: removed
  - Result.Items: removed
  - Result.Name: removed
  - ResultItem: removed
```

So that is not a reason for we block this PR.

File: go.mod (hunk 1 above)
Comment (camilamacedo86) [reply go.mod:15]: Yes, I tried it locally. 
The issue raised in the issue  https://github.com/kubernetes-sigs/kubebuilder/issues/2567 is not solved with the changes in this PR. See; https://github.com/kubernetes-sigs/kubebuilder/issues/2567#issuecomment-1086630767

**Comment by AlmogBaku** [comment]
@camilamacedo86 I changed it, although I disagree:

1. Fixing the problem with the go module should also fix #2567 
2. Upgrading the dependencies is another side effect we have there.

I think this PR is the combination of the upgrade and the bug.

**Comment by AlmogBaku** [comment]
@camilamacedo86 done

File: go.mod (hunk 1 above)
Comment (AlmogBaku) [reply go.mod:15]: Ok, my bad :)

Thanks for correcting me

**Pull request merged**